### PR TITLE
quality: parameterize sdk and test data branches

### PR DESCRIPTION
**Pull request by typotter**
🎟️ Fixes FF-3560 towards FF-3085

👯‍♂️ **Related PRs**

- [sdk-test-data](https://github.com/Eppo-exp/sdk-test-data/pull/83)
- [Update testing cloud function to accept branch name](https://github.com/Eppo-exp/eppo/pull/11418)

### Motivation
Changes are often made to the [sdk-test-data](https://github.com/Eppo-exp/sdk-test-data) repository to capture new behaviours, bugs and edge cases. When these changes are pushed to `main`, the SDKs are cloned locally (locally to the github action running) and their respective tests are run. These tests are set up and run by copies of the SDK test workflows - see [sdk-test-data workflow](https://github.com/Eppo-exp/sdk-test-data/blob/main/.github/workflows/test-sdks.yml). There are a number of limitations to this setup:

- Test steps are copied from the SDK’s respective workflows; changes to the SDK test workflows need to be replicated in sdk-test-data and this is not obvious to devs
- The SDK tests are not able to run against in-flight changes to `sdk-test-data`
- When new test-data is committed that breaks an SDK, that breakage is not surfaced in the SDK’s repository until some action triggers its testing workflow (on demand, on pull-request, etc.).

### Description of Changes
_This change_
🚀 - Each SDK's testing workflow is enhanced into a [reusable workflow](https://docs.github.com/en/actions/sharing-automations/reusing-workflows), exposing parameters for SDK branch and the sdk-test-data branch to use in testing.
- The test workflow runs using the main branch of sdk-test-data on all main pushes and Pull Requests using the PR's branch

_External to this Change_
[sdk-test-data](https://github.com/Eppo-exp/sdk-test-data/blob/main/.github/workflows) has two testing workflows.
1. ♻️  "Local Testing"- For all pull request changes, the "Local Testing" workflow calls the reusable SDK workflows, test results are recorded only in the sdk-test-data action. This is run using the main SDK branch and the "current" branch of workflow, i.e. the pull request branch. (SDK repo does not see/is not notified of test failures during PR lifecycle, only on push to main)
2. ♻️ 🧑‍💻 "Remote Testing" - On all pushes to the main branch of sdk-test-data, the SDK testing workflows are triggered to run within their respective repositories, alerting all subscribers of failed runs (repo is red/green).

### CODE REVIEW

File: .github/workflows/lint-test-sdk.yml [hunk 1]
```diff
@@ -5,12 +5,28 @@ on:
     paths:
       - '**/*'
   push:
-    branches: [main]
+    branches: [main, tp/**]
+
+  workflow_dispatch:
+
+  workflow_call:
```
Comment (typotter) [inline .github/workflows/lint-test-sdk.yml:12]: Allows specifying branches when used as a _reusable_ workflow

File: .github/workflows/lint-test-sdk.yml [hunk 2]
```diff
@@ -5,12 +5,28 @@ on:
     paths:
       - '**/*'
   push:
-    branches: [main]
+    branches: [main, tp/**]
+
+  workflow_dispatch:
```
Comment (typotter) [inline .github/workflows/lint-test-sdk.yml:10]: Used to allow dispatching via web API (remotely triggered test run when changes are made to `sdk-test-data`)

File: .github/workflows/lint-test-sdk.yml [hunk 3]
```diff
@@ -21,6 +37,8 @@ jobs:
     steps:
       - uses: actions/checkout@v4
         with:
+          repository: Eppo-exp/sdk-common-jdk
```
Comment (typotter) [inline .github/workflows/lint-test-sdk.yml:40]: explicitly specify the repo and ref otherwise, when run as a reusable workflow in another repo, this would not checkout the java-common SDK as required

File: Makefile [hunk 4]
```diff
@@ -40,7 +40,7 @@ test-data:
 	mkdir -p ${tempDir}
 	git clone -b ${branchName} --depth 1 --single-branch ${githubRepoLink} ${gitDataDir}
 	cp -r ${gitDataDir}/ufc ${testDataDir}
-	rm ${testDataDir}/ufc/bandit-tests/*.dynamic-typing.json
+	rm -f ${testDataDir}/ufc/bandit-tests/*.dynamic-typing.json || true
```
Comment (typotter) [inline Makefile:43]: bumped into this failing when I was doing some local testing that involved temporarily deleting most of the test data/cases. We shouldn't bump over this in any real scenario but no need to have this fail at any rate

File: src/test/java/cloud/eppo/BaseEppoClientTest.java [hunk 5]
```diff
@@ -39,8 +39,14 @@
 public class BaseEppoClientTest {
   private static final Logger log = LoggerFactory.getLogger(BaseEppoClientTest.class);
   private static final String DUMMY_FLAG_API_KEY = "dummy-flags-api-key"; // Will load flags-v1
-  private static final String TEST_HOST =
+
+  // Use branch if specified by env variable `TEST_DATA_BRANCH`.
+  private static final String TEST_BRANCH = System.getenv("TEST_DATA_BRANCH");
+  private static final String TEST_HOST_BASE =
       "https://us-central1-eppo-qa.cloudfunctions.net/serveGitHubRacTestFile";
+  private static final String TEST_HOST =
```
Comment (typotter) [inline src/test/java/cloud/eppo/BaseEppoClientTest.java:47]: see: https://github.com/Eppo-exp/eppo/pull/11418

File: .github/workflows/lint-test-sdk.yml [hunk 6]
```diff
@@ -32,4 +50,4 @@ jobs:
           gpg-passphrase: ${{ secrets.GPG_PASSPHRASE }}
 
       - name: Run tests
-        run: make test
+        run: make test branchName=${TEST_DATA_BRANCH}
```
Comment (aarsilv) [inline .github/workflows/lint-test-sdk.yml:53]: 🔥

**Review by aarsilv — APPROVED** [review]
🚀
